recipes/lora_dpo_single_device.py

```diff
@@ -0,0 +1,629 @@
+        the `torch.cuda.is_bf16_supported` API. This is disabled by default but can be enabled via
+        setting `dtype=bf16` in configuration.
+        - Checkpointing: of LoRA adapter parameters and their optimizer states. When resuming
+            from a checkpoint, the adapter parameters are loaded from the checkpoint along
+            with the base model weights. Note that intra-epoch resumption is not supported.
+        - Logging to terminal, WandB, or TensorBoard.
+
+    Assumptions:
+        - Checkpoints are ONLY saved at epoch boundaries. In case of failure, work done
+            in ongoing epoch is lost.
+        - Datasets are Map-style and data fits in memory (not streamed).
+
+    The following configs can be used to run this recipe:
+        >>> tune ls
+        RECIPE                          CONFIG
+        lora_finetune_single_device     llama2/7B_lora_single_device
+                                        llama2/7B_qlora_single_device
+
+    Args:
+        cfg (DictConfig): OmegaConf object parsed from yaml file
+
+    Raises:
+        ValueError: If ``dtype`` is set to fp16.
+        RuntimeError: If ``dtype`` is set to bf16 and the hardware does not support bf16.
+
+    """
+
+    def __init__(self, cfg: DictConfig) -> None:
+
+        self._device = utils.get_device(device=cfg.device)
+        # Reduced precision logic
+        self._dtype = utils.get_dtype(cfg.dtype, device=self._device)
+        # fp16 precision is explicitly disabled as it is not supported in this
+        # recipe (for example, no gradient scaling).
+        if self._dtype == torch.float16:
+            raise ValueError(
+                "fp16 precision is not supported in this recipe. Please use fp32 or bf16."
+            )
+        # For CUDA devices, check if the HW supports bf16 if bf16 is specified.
+        if (
+            self._dtype == torch.bfloat16
+            and self._device != torch.device("cpu")
+            and not torch.cuda.is_bf16_supported()
+        ):
+            raise RuntimeError("Full bf16 training is not supported on this hardware.")
+        # logging attributes
+        self._output_dir = cfg.output_dir
+        self._log_every_n_steps = cfg.log_every_n_steps if cfg.log_every_n_steps else 1
+        self._log_peak_memory_every_n_steps = 100
+
+        # These are public properties which are updated by the checkpoint loader
+        # when ``resume_from_checkpoint`` is `True` or validated in tests
+        self.seed = utils.set_seed(seed=cfg.seed)
+        self.epochs_run = 0
+        self.total_epochs = cfg.epochs
+        self.max_steps_per_epoch = cfg.max_steps_per_epoch
+        self.total_training_steps = 0
+
+        self._resume_from_checkpoint = cfg.resume_from_checkpoint
+        self._gradient_accumulation_steps = cfg.gradient_accumulation_steps
+
+    def load_checkpoint(self, cfg: DictConfig) -> Dict[str, Any]:
```

Comment:
nit: this should be updated in our other recipes, but let's rename this so we can distinguish between the checkpointer `DictConfig` object and the whole config `DictConfig` object
```suggestion
    def load_checkpoint(self, cfg_checkpointer: DictConfig) -> Dict[str, Any]:
```